phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexScrutinyMapper.java

```diff
@@ -254,6 +292,41 @@ private void processBatch(Context context)
         }
     }
 
+    private void checkIfInvalidRowsExpired(Context context,
+            Map<String, Pair<Long,
+            List<Object>>> targetPkToSourceValues) throws SQLException, IOException {
+        long ttl = getTableTtl();
+        Set<Map.Entry<String, Pair<Long, List<Object>>>> entrySet = targetPkToSourceValues.entrySet();
+
+        Iterator<Map.Entry<String, Pair<Long, List<Object>>>> itr = entrySet.iterator();
+
+        // iterate and remove items simultaneously
+        while(itr.hasNext()) {
+            Map.Entry<String, Pair<Long, List<Object>>> entry = itr.next();
+            Pair<Long, List<Object>> sourceValues = entry.getValue();
+            Long sourceTS = sourceValues.getFirst();
+            if (hasRowExpiredOnSource(sourceTS, ttl)) {
+                context.getCounter(PhoenixScrutinyJobCounters.EXPIRED_ROW_COUNT).increment(1);
+                itr.remove();
+            }
+        }
+    }
+
+    private boolean hasRowExpiredOnSource(Long sourceTS, Long ttl) {
+        long currentTS = EnvironmentEdgeManager.currentTimeMillis();
+        return ttl != Integer.MAX_VALUE && sourceTS + ttl*1000 < currentTS;
+    }
+
+    private long getTableTtl() throws SQLException, IOException {
+        PTable psourceTable = PhoenixRuntime.getTable(connection, qSourceTable);
+        Admin admin = connection.unwrap(PhoenixConnection.class).getQueryServices().getAdmin();
+        String fullTableName = SchemaUtil.getQualifiedTableName(psourceTable.getSchemaName().toString(),
+                psourceTable.getTableName().toString());
+        HTableDescriptor tableDesc = admin.getTableDescriptor(TableName.valueOf(fullTableName));
```

Comment:
If we have a lot of invalid rows, that's a lot of calls to getTableDescriptor, which sends an RPC to the HMaster. Is caching appropriate here? ("Weird stuff happens to your scrutiny run results if you change TTL during it" is an OK limitation, I think, since scrutiny mappers would probably fail during the TTL change anyway.)

Reply:
@swaroopak  Hmm...turns out I also need a frequent TableDescriptor lookup over at PHOENIX-5645 to verify that scn doesn't overshoot the ttl. Wonder if there's a need for a shared cache, similar to the PTable cache.

Reply:
In this case I can call getTableTtl in setup, isnt it?